src-built-in/components/UIAPI/effects/linker.ts

```diff
@@ -0,0 +1,158 @@
+ * @param isActive Whether the channel is active or not before the toggle
+ * @param windowIdentifier The window identifier for the current component the linker window is attached to
+ */
+export const linkChannel = (channelName: string, isActive: boolean, windowIdentifier: object, actives: number) => {
+    // Limit the number of maximum active channels to 6
+    if (actives > 5 && !isActive) {
+        return Promise.reject("The number of active channels shall not exceed 6.");
+    }
+    else {
+        return new Promise((res, rej) => {
+            const callback = (err: any, data: linkChannelReturnObject) => {
+                if (err) {
+                    FSBL.Clients.Logger.system.error(`Failed to togger channel ${channelName}: ${err}`);
+                    return rej(err);
+                }
+                FSBL.Clients.Logger.system.log(`Toggle channel success. Channel name: ${channelName}`);
+                const numActiveChannels = data.channels.length;
+                store.dispatch(updateActives(numActiveChannels));
+                res(data);
+            };
+            if (!isActive) {
+                FSBL.Clients.LinkerClient.linkToChannel(channelName, windowIdentifier, callback);
+            } else {
+                FSBL.Clients.LinkerClient.unlinkFromChannel(channelName, windowIdentifier, callback);
+            }
+            hideWindow();
+        });
+    }
+}
+
+// Window event listeners
+const hideWindow = () => {
+    finsembleWindow.hide();
+};
+const showWindow = () => {
+    finsembleWindow.focus();
+};
+
+/**
+ * Initialize the linker, which includes populating the linker state with the linker information (i.e. the number of channels, channel names,
+ * channel colors), setting up the responder when the linker state changes, and configuring the linker state according to finsemble config.
+ * @param initialState The initial state of the linker
+ */
+export const initializeLinker = (initialState: Linker) => {
+    FSBL.Clients.Logger.system.log(`Initializing the linker state. Current linker state: ${initialState}`);
+    finsembleWindow.addEventListener("blurred", hideWindow);
+    finsembleWindow.addEventListener("shown", showWindow);
+    
+    let nextChannelId: number = 0;
+    const initialLinkerState: Linker = Object.assign({}, initialState);
+    const initialChannels: Channels = {};
+    const initialNametoId: NameToId = {};
+
+    // Add the channels user specifies in the config to the linker's initial state
+    const addChanneltoInitialState = (channel: Channel) => {
+        initialChannels[nextChannelId] = {
+            id: nextChannelId,
+            name: channel.name,
+            color: channel.color,
+            border: channel.border,
+            active: false
+        };
+        initialNametoId[channel.name] = nextChannelId;
+        nextChannelId = nextChannelId + 1;
+    }
+
+    const updateChannels = (activeChannels: Array<Channel>) => {
```

Comment:
We will only be able to write unit tests for those methods if you move them out of the function body.